open-craft/opencraft
# Allow for longer PR branch names and mitigate GitHub rate limit issues

This pull request has two main goals: allow for automatic sandbox creation for longer branch names (255 characters up from the current 50 character limit) and also mitigates issues due to GitHub rate limiting.

To accomplish the second goal, we check the GitHub ratelimit status endpoint before making each call, and fail early if we don't have the resources available. We also deduplicate the requests that search for open pull requests by checking for PRs from multiple usernames at once.

**Dependencies**: None

**Testing instructions**:

1. Create a branch on your Ocim devstack's watched repo that has an excessively long name (50 characters or longer)
2. In the shell for your Ocim devstack, run the following:
    ```python
    >>> from pr_watch.tasks import watch_pr
    >>> watch_pr()
    ```
3. Verify that a new PR object is created without any issues.
4. Also in your Ocim devstack shell, run the following:
    ```python
    >>> from pr_watch.tasks import watch_pr
    >>> for i in range(100):
            watch_pr()
    ```
5. Watch the output, specifically eyeing the ratelimit HTTP responses, and note that once the "remaining" calls for search resources reach 0, you start seeing warnings that look like the following:
    ```
    Rate limit exceeded when requesting a 'search' resource at https://api.github.com/search/issues?sort=created&q=is:open is:pr author:opencraft-fake-account repo:random-fake-organization/edx-platform
    ```

**Author notes and concerns**:

1. This pull request shifts from calling the GitHub API once for every user on a team to calling it for all users on a team simultaneously by passing multiple `author:username` parameters.
2. We currently don't do anything when a ratelimited request fails; this is because every action we take (i.e., spawning new sandboxes) is logged in the database, so if a request fails, then we can simply pick up where we were at the next iteration.
